src/network/network_gui.cpp

```diff
@@ -1806,166 +1793,543 @@ static void PopupClientList(ClientID client_id, int x, int y)
 
-		this->DrawWidgets();
+				this->query_widget = WID_CL_SERVER_NAME_EDIT;
+				SetDParamStr(0, _settings_client.network.server_name);
+				ShowQueryString(STR_JUST_RAW_STRING, STR_NETWORK_CLIENT_LIST_SERVER_NAME_QUERY_CAPTION, NETWORK_NAME_LENGTH, this, CS_ALPHANUMERAL, QSF_ENABLE_DEFAULT | QSF_LEN_IN_CHARS);
+				break;
+
+			case WID_CL_CLIENT_NAME_EDIT:
+				this->query_widget = WID_CL_CLIENT_NAME_EDIT;
+				SetDParamStr(0, _settings_client.network.client_name);
+				ShowQueryString(STR_JUST_RAW_STRING, STR_NETWORK_CLIENT_LIST_PLAYER_NAME_QUERY_CAPTION, NETWORK_CLIENT_NAME_LENGTH, this, CS_ALPHANUMERAL, QSF_ENABLE_DEFAULT | QSF_LEN_IN_CHARS);
+				break;
+
+			case WID_CL_SERVER_VISIBILITY:
+				if (!_network_server) break;
+
+				ShowDropDownMenu(this, _server_visibility_dropdown, _settings_client.network.server_advertise, WID_CL_SERVER_VISIBILITY, 0, 0);
+				break;
+
+			case WID_CL_MATRIX:
+				uint index = this->vscroll->GetScrolledRowFromWidget(pt.y, this, WID_CL_MATRIX);
+				NWidgetBase *widget = this->GetWidget<NWidgetBase>(WID_CL_MATRIX);
+
+				bool rtl = _current_text_dir == TD_RTL;
+				uint x = rtl ? (uint)WD_FRAMERECT_LEFT : widget->current_x - WD_FRAMERECT_RIGHT;
+
+				/* Find the buttons for this row. */
+				auto button_find = this->buttons.find(index);
+				if (button_find == this->buttons.end()) break;
+
+				/* Check if we clicked on any of the buttons. */
+				for (auto &button : button_find->second) {
+					uint left = rtl ? x : x - button.width;
+					uint right = rtl ? x + button.width : x;
+
+					if (IsInsideMM(pt.x, left, right)) {
+						if (button.company_proc != nullptr) {
+							button.company_proc(this, pt, button.company_id);
+						}
+						if (button.client_proc != nullptr) {
+							button.client_proc(this, pt, button.client_id);
+						}
+					}
+
+					uint width = button.width + WD_FRAMERECT_LEFT + WD_FRAMERECT_RIGHT;
+					x += rtl ? width : -width;
+				}
+
+				break;
+		}
 	}
 
-	void DrawWidget(const Rect &r, int widget) const override
+	bool OnTooltip(Point pt, int widget, TooltipCloseCondition close_cond) override
 	{
-		if (widget != WID_CL_PANEL) return;
+		switch (widget) {
+			case WID_CL_SERVER_NAME:
+				GuiShowTooltips(this, STR_NETWORK_CLIENT_LIST_SERVER_NAME_TOOLTIP, 0, nullptr, close_cond);
+				return true;
 
-		bool rtl = _current_text_dir == TD_RTL;
-		int icon_offset = (this->line_height - icon_size.height) / 2;
-		int text_offset = (this->line_height - FONT_HEIGHT_NORMAL) / 2;
+			case WID_CL_CLIENT_NAME:
+				GuiShowTooltips(this, STR_NETWORK_CLIENT_LIST_PLAYER_NAME_TOOLTIP, 0, nullptr, close_cond);
+				return true;
 
-		uint y = r.top + WD_FRAMERECT_TOP;
-		uint left = r.left + WD_FRAMERECT_LEFT;
-		uint right = r.right - WD_FRAMERECT_RIGHT;
-		uint type_icon_width = this->server_client_width + this->icon_size.width + WD_FRAMERECT_LEFT;
+			case WID_CL_MATRIX: {
+				uint index = this->vscroll->GetScrolledRowFromWidget(pt.y, this, WID_CL_MATRIX);
```

Comment:
This code looks awfully similar to the last case in the function above. Since I would reckon that clicking can also only happen on a single button at a time, maybe put all the logic in a `Button *GetButtonAtPoint(Point pt)` that returns the Button, and then based on whether it is `nullptr` or not you either call the OnClick or GuiShowTooltips function.